tool/tsh/common/tsh.go

```diff
@@ -3558,15 +3558,16 @@ func makeClientForProxy(cf *CLIConf, proxy string) (*client.TeleportClient, erro
 
 	// If we are missing client profile information, ping the webproxy
 	// for proxy info and load it into the client config.
-	if profileError != nil || cf.IdentityFileIn != "" {
+	if profileError != nil || profile.MissingClusterDetails {
 		log.Debug("Pinging the proxy to fetch listening addresses for non-web ports.")
 		_, err := tc.Ping(cf.Context)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
-		// Identityfile uses a placeholder profile. Save missing profile info.
-		if cf.IdentityFileIn != "" {
+		// This is a placeholder profile created from limited cluster details.
+		// Save missing cluster details gathererd during Ping.
+		if profileError == nil && profile.MissingClusterDetails {
```

Comment:
Missing cluster details is a new flag to say the profile is incomplete. With only the identity we don't have all the information needed to build a complete profile. We need the ping's answer to know the web URL and how's the cluster must be contacted.

The only time when `profile.MissingClusterDetails` is true is when the profile is created from an identity. This is equivalent to `cf.IdentityFileIn != ""` but much more robust as one can clear the CLI flags and the profile will still contain the right information.

This change doesn't affect the existing login flow and is backward compatible.

Reply:
Ok, so `profile.MissingClusterDetails` is basically a replacement for `if cf.IdentityFileIn != ""`.

What about `profileError == nil` condition here? This condition is not equivalent to the previous one: before we would always call `tc.SaveProfile()` regardless of whether `profileError` is nil or not, now we require it to be nil. Is this not a change in behavior?

Reply:
And it's not really a change in behavior. We always try to parse an identity file into a profile with missing cluster details. If we failed to make the profile, we would fail before here, so profileError would always be nil here if an identity file was in use. We just need to do this check explicitly now because if `profileError != nil` then `profile == nil` and `profile.MissingClusterDetails` results in a nil panic.